internal/overlord/logstate/forwarder.go

```diff
@@ -0,0 +1,232 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	parserSize = 4 * 1024
+
+	maxBufferedLogs = 100
+)
+
+var flushDelay = 1 * time.Second
+
+// logForwarder is responsible for forwarding logs from a single service to
+// a single target.
+type logForwarder struct {
+	service string
+	target  *plan.LogTarget
+
+	iterator   servicelog.Iterator
+	parser     *servicelog.Parser
+	iteratorCh chan struct{}
+	stopCh     chan chan struct{}
+
+	messages    []servicelog.Entry
+	maxMessages int
+
+	delay      time.Duration
+	flushTimer *time.Timer
+	timerSet   bool
+	timeoutCh  chan struct{}
+
+	client logClient
+}
+
+func newLogForwarder(
+	service string, target *plan.LogTarget, iterator servicelog.Iterator,
+) (*logForwarder, error) {
+
+	var client logClient
+	var err error
+
+	switch target.Type {
+	case plan.LokiTarget:
+		client = newLokiClient(target.Location)
+
+	case plan.SyslogTarget:
+		client, err = newSyslogClient(target.Location)
+
+	default:
+		return nil, fmt.Errorf(
+			"unknown log target type %q", target.Type)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return newLogForwarderInternal(service, target, iterator, client, maxBufferedLogs, flushDelay), nil
+}
+
+// newLogForwarderInternal is shared by the real and test code.
+func newLogForwarderInternal(
+	service string, target *plan.LogTarget, iterator servicelog.Iterator,
+	client logClient, maxMessages int, delay time.Duration,
+) *logForwarder {
+
+	iteratorCh := make(chan struct{})
+	timeoutCh := make(chan struct{}, 1)
+	// Create timer. Here we set the timer duration to a large value and
+	// immediately stop it, to ensure the timer starts off unset.
+	flushTimer := time.AfterFunc(time.Hour, func() {
+		timeoutCh <- struct{}{}
+		// If we are waiting on iterator.Next, cancel this
+		select {
+		case iteratorCh <- struct{}{}:
+		default:
+		}
+	})
+	flushTimer.Stop()
+
+	return &logForwarder{
+		service:     service,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, parserSize),
+		iteratorCh:  iteratorCh,
+		stopCh:      make(chan chan struct{}, 1),
+		maxMessages: maxMessages, // TODO: make this value configurable in plan
+		delay:       delay,       // TODO: make this value configurable in plan
+		flushTimer:  flushTimer,
+		timerSet:    false,
+		timeoutCh:   timeoutCh,
+		client:      client,
+	}
+}
+
+// forward continually pulls logs from the service buffer and forwards them
+// to the remote log target.
+// This method will block until the service is stopped, or stop() is called.
+// Hence, it should be run in a separate goroutine.
+func (f *logForwarder) forward() {
```

Comment:
We've discussed possibly ways to simplify this via video. [Pastebin for the record.](https://pastebin.canonical.com/p/QQ5xHrw6yg/)